gpcontrib/gp_orca/gporca.c

```diff
@@ -1,14 +1,120 @@
 #include "postgres.h"
 
 #include "cdb/cdbvars.h"
+#include "executor/executor.h"
 #include "miscadmin.h"
+#include "optimizer/orca.h"
+#include "optimizer/planner.h"
+#include "storage/ipc.h"
 #include "utils/builtins.h"
+#include "utils/guc.h"
+#include "utils/memutils.h"
 
 PG_MODULE_MAGIC;
 
+extern void InitGPOPT();
+extern void TerminateGPOPT();
+
+extern void compute_jit_flags(PlannedStmt* pstmt);
+
 void		_PG_init(void);
 void		_PG_fini(void);
 
+static planner_hook_type prev_planner = NULL;
+
+static void
+gp_orca_shutdown(int code, Datum arg)
+{
+	TerminateGPOPT();
+
+	if (OptimizerMemoryContext != NULL)
+		MemoryContextDelete(OptimizerMemoryContext);
+}
+
+static void
+gp_orca_init()
+{
+	/* Initialize GPOPT */
+	OptimizerMemoryContext = AllocSetContextCreate(TopMemoryContext,
+												"GPORCA Top-level Memory Context",
+												ALLOCSET_DEFAULT_MINSIZE,
+												ALLOCSET_DEFAULT_INITSIZE,
+												ALLOCSET_DEFAULT_MAXSIZE);
+
+	InitGPOPT();
+
+	before_shmem_exit(gp_orca_shutdown, 0);
+}
+
+static PlannedStmt *
+gp_orca_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
+{
+	PlannedStmt	   *result = NULL;
+	static bool first_call = true;
+
+	/*
+	 * Use ORCA only if it is enabled and we are in a coordinator QD process.
+	 *
+	 * ORCA excels in complex queries, most of which will access distributed
+	 * tables. We can't run such queries from the segments slices anyway because
+	 * they require dispatching a query within another - which is not allowed in
+	 * GPDB (see querytree_safe_for_qe()). Note that this restriction also
+	 * applies to non-QD coordinator slices.  Furthermore, ORCA doesn't currently
+	 * support pl/<lang> statements (relevant when they are planned on the segments).
+	 * For these reasons, restrict to using ORCA on the coordinator QD processes only.
+	 *
+	 * PARALLEL RETRIEVE CURSOR is not supported by ORCA yet.
+	 */
+	if (optimizer &&
+		GP_ROLE_DISPATCH == Gp_role &&
+		IS_QUERY_DISPATCHER() &&
+		(cursorOptions & CURSOR_OPT_SKIP_FOREIGN_PARTITIONS) == 0 &&
+		(cursorOptions & CURSOR_OPT_PARALLEL_RETRIEVE) == 0)
+	{
+		instr_time		starttime;
+		instr_time		endtime;
+
+		if (first_call)
+		{
+			gp_orca_init();
+			first_call = false;
+		}
```

Comment:
how about something like
```suggestion
		if (OptimizerMemoryContext == NULL)
			gp_orca_init();
```